Limit message sizes before receiving.

sipa opened the pull request:
There is currently no message over 2 MiB that is acceptable anyway, no need to support them inside our network buffers.

This is done by a dispatch mechanism in net, rather than a direct check, to support future modularity (where message handling is split over different modules). It's also required to deal with DoS management correctly (which is done outside of net).

It also doesn't touch serialize.h's MAX_SIZE, as it's a global for all serialization usage, including non-protocol processing.

gmaxwell commented:
ACK.  Also tested by IBDing a node running this against a node running this.

laanwj commented:
Code looks correct to me, going to test.

laanwj commented on src/main.cpp:
```diff
@@ -4253,6 +4258,18 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
+    if (msg->nMessageSize > MAX_PROTOCOL_MESSAGE_LENGTH) {
+        LOCK(cs_main);
+        Misbehaving(pnode->GetId(), 100);
+        LogPrintf("net", "Oversized message from peer=%d (size=%d)\n", pnode->GetId(), msg->nMessageSize);
```
I lobbed a 4MB message at my node and it crashed here:
```
EXCEPTION: St13runtime_error       
tinyformat: Not enough conversion specifiers in format string       
bitcoin in net       
```
As a category is passed, this should be `LogPrint`, not `LogPrintf`.

sipa replied:
Fixed.

gmaxwell commented:
Retesting.

laanwj commented:
Tested ACK
```
2015-03-03 19:19:20 initial getheaders (324870) to peer=1 (startheight:0)
2015-03-03 19:19:20 sending: getheaders (997 bytes) peer=1
2015-03-03 19:19:20 Misbehaving: 127.0.0.1:57403 (0 -> 100) BAN THRESHOLD EXCEEDED
2015-03-03 19:19:20 Oversized message from peer=1 (size=4194304)
2015-03-03 19:19:20 socket closed
2015-03-03 19:19:20 disconnecting peer=1
```

sipa committed:
Limit message sizes before transfer

This introduces a fixed limit for the size of p2p messages, and enforces it
before download.

sipa commented:
@gmaxwell Please test again?

gmaxwell commented:
It disconnects now, but as I pointed out-- while it grants misbehaving-- it does not attempt to ban because now it disconnects before it gets a chance to do so. If it's not going to ban, there is no point in setting the misbehaving, and a simpler patch which doesn't use the signals would work the same. (though I suppose we should have the combiner fix for the other signal that uses the return value!).

gavinandresen commented:
ACK on the signal combiner.

RE: Misbehaving/banning: seems like it should ban, but instead of introducing Yet Another Signal, seems to me a check for state->fShouldBan in main.cpp FinalizeNode (called when net disconnects a peer) might be the right approach.

gavinandresen commented:
@sipa : nit:  I think the combiner code could be replaced with boost::bind(boost::algorithm::all_of_equal, _1, _2, true)

sipa commented:
Moved the combiner logic, and simplified the code here.

A more complex signals-based solution can be done later.

gmaxwell commented:
I was a little surprised that you went with logging at only debug=net level, but it works. ACK.

laanwj commented:
I like the simpler solution. ACK.

jgarzik commented:
posthumous ACK for simpler solution
